src/sources/soundsourcestem.cpp

```diff
@@ -0,0 +1,468 @@
+        // No special channel format request
+        initChannelCountOnce(
+                mixxx::audio::ChannelCount::stereo() *
+                m_pStereoStreams.size());
+    }
+
+    initSampleRateOnce(m_pStereoStreams.front()->getSignalInfo().getSampleRate());
+    initBitrateOnce(m_pStereoStreams.front()->getBitrate());
+    initFrameIndexRangeOnce(m_pStereoStreams.front()->frameIndexRange());
+
+    return OpenResult::Succeeded;
+}
+
+void SoundSourceSTEM::close() {
+    for (auto& stream : m_pStereoStreams) {
+        stream->close();
+    }
+}
+
+ReadableSampleFrames SoundSourceSTEM::readSampleFramesClamped(
+        const WritableSampleFrames& globalSampleFrames) {
+    if (m_pStereoStreams.size() == 1) {
+        // Opened in stereo mode, proxy the read request to the main mix channel
+        return m_pStereoStreams.front()->readSampleFrames(globalSampleFrames);
+    }
+
+    int stemCount = m_pStereoStreams.size();
+
+    VERIFY_OR_DEBUG_ASSERT(globalSampleFrames.writableLength() %
+                    (stemCount * mixxx::audio::ChannelCount::stereo()) ==
+            0) {
+        return ReadableSampleFrames();
+    };
+
+    SINT stemSampleLength = m_pStereoStreams.front()->getSignalInfo().frames2samples(
+            globalSampleFrames.frameLength());
+
+    // The same buffer is reused between requests tp prevent reallocation, but
+    // it will be reallocated if a larger chunk is requested and will keep the
+    // new maximum size
+    if (stemSampleLength > m_buffer.size()) {
+        m_buffer = SampleBuffer(stemSampleLength);
+    }
+
+    ReadableSampleFrames read(globalSampleFrames.frameIndexRange(),
+            SampleBuffer::ReadableSlice(
+                    globalSampleFrames.writableData(),
+                    globalSampleFrames.writableLength()));
+    DEBUG_ASSERT(stemSampleLength * stemCount == globalSampleFrames.writableLength());
+    CSAMPLE* pBuffer = globalSampleFrames.writableData();
+    for (int streamIdx = 0; streamIdx < stemCount; streamIdx++) {
+        WritableSampleFrames currentStemFrame = WritableSampleFrames(
+                globalSampleFrames.frameIndexRange(),
+                SampleBuffer::WritableSlice(
+                        m_buffer.data(),
+                        stemSampleLength));
+        m_pStereoStreams[streamIdx]->readSampleFrames(currentStemFrame);
+        // TODO(XXX): currently, stem samples are interleaved and packed next to each other as such:
+        //    1L1R1L1R1L1R...2L2R2L2R2L2R2L2R......3L3R3L3R3L3R3L3R......4L4R4L4R4L4R4L4R....
+        //    Can FFmpeg decode as without having to use a decoder per channel?
+        //    1LLLLLLLLLLLLLL....1RRRRRRRRR...2LLLLLLL...?
+
+        // Change the sample layout to interleave all channels together
+        for (SINT i = 0; i < stemSampleLength / 2; i++) {
+            pBuffer[2 * stemCount * i + 2 * streamIdx] = m_buffer[2 * i];
+            pBuffer[2 * stemCount * i + 2 * streamIdx + 1] = m_buffer[2 * i + 1];
+        }
```

Comment:
I assume that this memory structure is to be optimized across all modules in follow-up PR?

Reply:
I guess by optimisation, you mean have FFmpeg take care of the layout when decoding? I don't have any immediate plan. I had a quick look and I didn't find any simple solutions for this as each stream is technically independent, handled by a dedicated FFmpeg context. Do you have any suggestion or tips on how to do so? 
FYI, I took some stats to see if that for-loop/re-layout was a bottleneck, and it is about 1/100th of the time used by RubberBand so probably not a problem yet.

Reply:
The rubberband performance is a huge problem - was already before, but stems multiplies this. On my rather fast Windows laptop, I can only use a singles stems deck with Rubberband R2 and R3 becomes unusable with stems.
I could imagine, that one major reason for this is, that the data are not alligned in the memory, to allow the compiler to use efficient SIMD instructions. Maybe this can only be changed in Rubberband itself, but maybe already the memory alignment in our engine might influence this.
Therefore I think, that analysis and overall optimization of the engines memory structures might be necessary.

Reply:
> but maybe already the memory alignment in our engine might influence this

This part is already aligned, but I assume you me the [channel buffer](https://github.com/acolombier/mixxx/blob/b8c19a0c2ba6b77971d2e15cf84faa258610e018/src/engine/bufferscalers/enginebufferscalerubberband.cpp#L166)

Are we sure RubberBand would use SIMD? If yes, I can try to scrape the individual `SampleBuffer` and use a a contiguous mem alloc, instead of individual ones

Reply:
(Note that I had plenty of `ScopedTimer` all across which allow me to confirm that 99% of the CPU time is due to RubberBand::retrieve, which I removed now)

Reply:
I did the experiment with a contiguous global buffer and there was no changes.
Just FYI, I've put back the `ScopedTimer`  and here is the result
```
Stat("EngineBufferScaleRubberBand::scaleBuffer","count=462,sum=2.82884e+09ns,average=6.12302e+06ns,min=2.77938e+06ns,max=1.97878e+07ns,variance=3.76359e+12ns^2,stddev=1.94e+06ns")
Stat("EngineDeck::processStem","count=535,sum=2.88642e+09ns,average=5.39517e+06ns,min=89170ns,max=2.01299e+07ns,variance=7.62824e+12ns^2,stddev=2.76193e+06ns")
Stat("RubberBand::process","count=1839,sum=2.81529e+09ns,average=1.53088e+06ns,min=7860ns,max=4.10194e+06ns,variance=8.39907e+10ns^2,stddev=289812ns")
Stat("RubberBand::retrieve","count=2299,sum=2.65024e+06ns,average=1152.78ns,min=105ns,max=16231ns,variance=1.78589e+06ns^2,stddev=1336.37ns")
```
(Just the sum)
```
EngineDeck::processStem                  2.88642e+09ns
EngineBufferScaleRubberBand::scaleBuffer 2.82884e+09ns
RubberBand::process                      2.81529e+09ns
RubberBand::retrieve                     2.65024e+06ns
```
It was `RubberBand::process` btw, not `RubberBand::retrieve`

I guess another experiment would be to aligned the buffer size with the [sample count](https://github.com/mixxxdj/mixxx/blob/b8c19a0c2ba6b77971d2e15cf84faa258610e018/src/engine/bufferscalers/enginebufferscalerubberband.cpp#L261), I'll try that tomorrow.

Reply:
Great, that you did the analysis already! It would be even greater, to store such results in a document.
A helpul tool to understand such performance issues is callgrind. I used it once (for a much smaller single thread project), and could clearly see, where the CPU cycles were burned. And I could see, where the compiler picked SIMD implementations of std library functions, and where not.